File: code/python/01.1.1_blink/blink.py
#!/usr/bin/env python3
##
## Filename:    blink.py
## Description: basic usage of GPIO. Let LED blink.
##
########################################################################################

########################################################################################
##
## import `LED` functionality from `gpiozero` library, and `sleep` from `time` lib
##
from gpiozero import LED
from time import sleep

########################################################################################
##
## LEDpin is our LED instance connected to GPIO BCM pin (pin 17)
##
## The following addressing modes are available to use (these are all equivalent):
##
##    LEDpin = LED(17)           # BCM
##    LEDpin = LED("GPIO17")     # BCM
##    LEDpin = LED("BCM17")      # BCM
##    LEDpin = LED("BOARD11")    # BOARD
##    LEDpin = LED("WPI0")       # WiringPi
##    LEDpin = LED("J8:11")      # BOARD
##
LEDpin = LED(17)

def loop():
    while True:

        ################################################################################
        ##
        ## turn on the LED
        ##
        LEDpin.on()

        ################################################################################
        ##
        ## update information displayed to terminal
        ##
        print ("LED is: ON")

        ################################################################################
        ##
        ## strategic delay for user perception (1 second)
        ##
        sleep (1)

        ################################################################################
        ##
        ## turn off the LED
        ##
        LEDpin.off()

        ################################################################################
        ##
        ## update information displayed to terminal
        ##
        print ("LED is: OFF")  # print message on terminal

        ################################################################################
        ##
        ## strategic delay for user perception (1 second)
        ##
        sleep (1)    # wait 1 second

########################################################################################
##
## start of program
##
if __name__ == '__main__':

    ####################################################################################
    ##
    ## display operational information
    ##
    print ("Program is starting (CTRL-c to interrupt) ...")
    print (f"LEDpin is connected to pin: {LEDpin}\n")

    ####################################################################################
    ##
    ## normal operations: run the loop
    ##
    try:
        loop()

    ####################################################################################
    ##
    ## catch any keyboard interrupt characters to signal program exit
    ##
    except KeyboardInterrupt:
        print ("ending program")

File: code/C++/01.1.1_blink/blink.cc
//
// Filename:    blink.c
// Description: libLED C++ test program (LED blinks every second)
// Requires:    libLED built/installed from the libraries/ directory
//
////////////////////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <wiringPi.h>
#include <LED.h>                                     // libLED-provided header

#define  LEDpin    0                                 // LED wiringPi pin number

int main (void) {

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Declare and initialize variables
    //
    LED *blink            = NULL;

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Bring wiringPi functionality online
    //
    if (wiringPiSetup () == -1)
    {
        std::cout << "[ERROR] Could not initialize wiringPi Library!" << std::endl;
        exit (1);
    }

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // create and initialize a new LED instance (called `blink`)
    //
    blink                 = new LED (LEDpin, LOW);

    fprintf (stdout, "Program is starting (CTRL-c to interrupt ...\n");
    fprintf (stdout, "[LED] configured for OUTPUT mode using pin: %d\n", blink.getPin ());

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Display the default status (note the lack of newline at the end, so that with
    // the strategic use of backspaces, we can update this line in place)
    //
    fprintf (stdout, "\nLED: OFF (%4s)", blink.getState ());

    while (1) {

        ////////////////////////////////////////////////////////////////////////////////
        //
        // Turn LED on
        //
        blink.write (HIGH);

        ////////////////////////////////////////////////////////////////////////////////
        //
        // update information display to report current LED state
        //
        fprintf (stdout, "\b\b\b\b\b\b\b\b\b\bON  (%4s)", blink.getState ());

        ////////////////////////////////////////////////////////////////////////////////
        //
        // strategic delay for user perception (1000mS == 1 second)
        //
        delay (1000);

        ////////////////////////////////////////////////////////////////////////////////
        //
        // Turn LED off
        //
        blink.write (LOW);

        ////////////////////////////////////////////////////////////////////////////////
        //
        // update information display to report current LED state
        //
        fprintf (stdout, "\b\b\b\b\b\b\b\b\b\bOFF (%4s)", blink.getState ());

        ////////////////////////////////////////////////////////////////////////////////
        //
        // strategic delay for user perception (1000mS == 1 second)
        //
        delay (1000);
    }

    return (0);
}

